### PR TITLE
Initial work of tls.connect and tls.write

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
Work in progress patch of TLS module integration with MbedTLS.

IoT.js-DCO-1.0-Signed-off-by: Zsolt Borbély zsborbely.u-szeged@partner.samsung.com
IoT.js-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

File: src/js/tls.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+Tls.TLSSocket.prototype.disableRenegotiation = function() {
+  this.disableRenegotiation = true;
+}
+
+Tls.TLSSocket.prototype.encrypted = function() {
+  return this.encrypted;
+}
+
+Tls.TLSSocket.prototype.getCipher = function() {
+}
```
Comment (LaszloLango) [inline src/js/tls.js]: What about throwing `new TypeError("Unimplemented");` in every unimplemented functions?

Comment (DanielBallaSZTE) [reply src/js/tls.js]: Great idea!

File: cmake/mbedtls.cmake [hunk 2]
```diff
@@ -14,17 +14,17 @@
 
 cmake_minimum_required(VERSION 2.8)
 
-if(NOT EXPERIMENTAL_BUILD_MBEDTLS)
-  return()
-endif()
-
 if(NOT "${TARGET_BOARD}" STREQUAL "None")
```
Comment (LaszloLango) [inline cmake/mbedtls.cmake]: We can remove this condition too. I've tested in RPi2 and it works well.

File: src/modules/iotjs_module_tls.c [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  const char* buf_orig =
+      iotjs_buffer_allocate(sizeof(char) * iotjs_string_size(&str));
+  buf_orig = iotjs_string_data(&str);
+  unsigned char buf[1024];
+  int len = sprintf((char*)buf, buf_orig);
+  iotjs_buffer_release((char*)buf_orig);
+  int ret = 0;
+  while ((ret = mbedtls_ssl_write(&tls_data->ssl, buf, (size_t)len)) <= 0) {
+    if (ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE) {
+      return create_error_message("write error, error code: %d", ret);
```
Comment (LaszloLango) [inline src/modules/iotjs_module_tls.c]: I think there is no need for `create_error_message`. I'd simply use `snprintf` and `JS_CREATE_ERROR`. Even if we keep the function, it must return with an error value.

File: src/modules/iotjs_module_tls.c [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  return jerry_create_string_sz((const jerry_char_t*)buff, err_len);
+}
+
+JS_FUNCTION(Write) {
+  DJS_CHECK_ARGS(1, string);
+  JS_DECLARE_THIS_PTR(tls, tls_data);
+  iotjs_string_t str = JS_GET_ARG(0, string);
+
+  const char* buf_orig =
+      iotjs_buffer_allocate(sizeof(char) * iotjs_string_size(&str));
```
Comment (hs0225) [inline src/modules/iotjs_module_tls.c]: The allocated buffer is not used. You assign another string in the code below. Therefore this line is not necessary.

File: src/modules/iotjs_module_tls.c [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  buf_orig = iotjs_string_data(&str);
+  unsigned char buf[1024];
+  int len = sprintf((char*)buf, buf_orig);
+  iotjs_buffer_release((char*)buf_orig);
```
Comment (hs0225) [inline src/modules/iotjs_module_tls.c]: Please change to `iotjs_string_destroy(&str);`.

File: src/modules/iotjs_module_tls.c [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+JS_FUNCTION(Write) {
+  DJS_CHECK_ARGS(1, string);
+  JS_DECLARE_THIS_PTR(tls, tls_data);
+  iotjs_string_t str = JS_GET_ARG(0, string);
+
+  const char* buf_orig =
+      iotjs_buffer_allocate(sizeof(char) * iotjs_string_size(&str));
+  buf_orig = iotjs_string_data(&str);
+  unsigned char buf[1024];
+  int len = sprintf((char*)buf, buf_orig);
```
Comment (hs0225) [inline src/modules/iotjs_module_tls.c]: You can use `iotjs_string_size(&str)` instead of it.

File: src/modules/iotjs_module_tls.c [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+  return jerry_create_number(ret);
+}
+
+
+JS_FUNCTION(Connect) {
+  DJS_CHECK_ARGS(3, string, string, string);
+  iotjs_string_t port = JS_GET_ARG(0, string);
+  iotjs_string_t host = JS_GET_ARG(1, string);
+  iotjs_string_t hostname = JS_GET_ARG(2, string);
```
Comment (hs0225) [inline src/modules/iotjs_module_tls.c]: You need to release the buffer with `iotjs_string_destroy` function about `port`, `host`, `hostname`.

**Comment by hs0225** [comment]
@DanielBallaSZTE Please correct the following two styles in JavaScript code correctly.

1. double quotes -> single quotes
```
throw new TypeError('Unimplemented');
```

2. The function expressions must end with `;`.
```js
Tls.Server = function(options) {
  return net.Server(options);
};
```

**Comment by DanielBallaSZTE** [comment]
@hs0225 @LaszloLango I have updated the pull request according to you reviews, thank you.

File: src/modules/iotjs_module_tls.c [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  buf = iotjs_buffer_allocate(sizeof(char) * 256);
+  vsnprintf(buf, sizeof(char) * 256, format, arg_list);
+  va_end(arg_list);
+  jerry_value_t ret_val = jerry_create_string((jerry_char_t*)buf);
+  iotjs_string_destroy(h);
+  iotjs_string_destroy(p);
+  iotjs_string_destroy(hn);
+  iotjs_buffer_release(buf);
+
+  return ret_val;
```
Comment (hs0225) [inline src/modules/iotjs_module_tls.c]: Please use `JS_CREATE_ERROR(TYPE, message)` to throw an error.

Comment (DanielBallaSZTE) [reply src/modules/iotjs_module_tls.c]: You are right! Thank you.

**Pull request merged**